### PR TITLE
fix(llm-utils): stop silently dropping array-content user messages with no image block

### LINKED ISSUE #1113: fix(llm): user messages with array content and no image are silently dropped from assembly
### Problem

`buildAndSortMessages` splits `fabMessages` into two sets and a `role: 'user'` message whose content is an array with no image-typed block matches neither, so it is silently omitted from the assembled payload:

- `nonImageMessages` requires `!Array.isArray(content)` (`b4m-core/utils/src/llm/utils.ts`)
- `imageMessages` requires `Array.isArray(content)` AND a block whose `type` starts with `image`

An array-content user message carrying only text, `tool_use`, or `tool_result` blocks falls through both filters. Pre-existing behaviour, not introduced by #1004; raised as a P3 while reviewing it.

Whether it is currently reachable is the open question. `processFabFilesServer` pushes its filename/text block into the same array as the image, so its array-content messages always carry an image. But the tool-pairing logic in the same function (`ensureToolPairingIntegrity`, and the `historyEndsWithToolUse` / `promptHasToolResult` handling) exists precisely because `tool_use` / `tool_result` messages occur, and any caller passing one through `fabMessages` would lose it without a trace.

### Fix

Make the split exhaustive rather than two positive filters: partition on "has an image block", so every user message lands in exactly one set. Anything not classified should be logged rather than dropped, on the principle that a silent drop whose symptom is a missing answer becomes a support ticket.

### Scope

- `b4m-core/utils/src/llm/utils.ts` - the two filters and the assembly order
- Ordering needs care: the tool_use/tool_result adjacency requirement means a recovered message cannot simply be appended

### Acceptance criteria

- A `role: 'user'` array-content message with only text blocks reaches the model
- A `tool_use` / `tool_result` array-content message either reaches the model in a position that preserves pairing, or is dropped with a warning that names what was dropped and why
- A test drives a text-only array-content fab message and asserts it is present in the output - it should fail against the current filters
- No change to image-carrying messages: the existing image tests still pass unmodified

### PR DESCRIPTION
**Pull request by cleffrem-dev**
# Pull Request

## Optional Screenshot/Video
live-extraContextMessages-proof-before

*Before (staging, main): `POST /api/ai/llm` with an array-content-only `extraContextMessages` entry carrying a passphrase. The model replies that it has no such passphrase, proving the message never reached it.*

live-extraContextMessages-proof-after

*After (this PR's live preview): the same request, same passphrase. The model echoes it back correctly, proving the fix delivers the message.*

## Description
`buildAndSortMessages` partitioned incoming context messages with two positive filters that left a gap: a `role: 'user'` message whose content is an array but carries no image block (only text, or a `tool_use`/`tool_result`/`thinking` block) matched neither filter and was dropped with no trace. This is reachable today via the `extraContextMessages` request field, which accepts arbitrary array content, so an accepted API input was being silently discarded.

Replaces the two filters with one exhaustive classification pass so every message lands in exactly one bucket. Text-only array content now goes through the same budget/truncation pipeline as string content. A `tool_use`/`tool_result`/`thinking`-bearing message can't be delivered safely from this argument (tool pairing is reconstructed from conversation history, not from this input), so those are dropped with a warning naming what was dropped and why, instead of vanishing silently. Also closes the same gap for any non-user/non-system message, and fixes a `.startsWith('image')` check that could throw on a malformed block.

Closes #1113

## Changes
- Replace the two-filter `nonImageMessages`/`imageMessages` split in `buildAndSortMessages` with one exhaustive classification pass (image / text-only / undeliverable-blocks / unassignable-role)
- Log a warning naming what was dropped and why, for both the tool/thinking-bearing case and any non-user/non-system message
- Guard `isAttachment` and `attachmentLabel` so array-content context blocks are never mislabeled or accounted as a file attachment
- Swap a throw-prone `.startsWith('image')` check for the existing null-safe `isImageBlock` helper
- Add 8 tests covering the new classification, the drop-with-warning path, and the existing image-priority/edge-case behavior

## Guide for Testers
The bug is reachable live via the `extraContextMessages` request field on `POST /api/ai/llm`. There is no UI affordance for it; only a programmatic caller (an API-key integration) would send this shape today.

1. Authenticate (bearer token) and call:
   ```
   POST /api/ai/llm
   {
     "message": "Reply with ONLY the archive passphrase mentioned in your context above, and nothing else.",
     "fabFileIds": [], "messageFileIds": [], "historyCount": 0,
     "params": { "model": "<any available text model, see GET /api/models>" },
     "extraContextMessages": [
       { "role": "user", "content": [{ "type": "text", "text": "Archive note: the archive passphrase is zamboni-quartz-19." }] }
     ]
   }
   ```
2. Poll `GET /api/quests/{id}` (the `quest.id` from the response above) until `replies` is non-empty.
3. On this PR: the reply contains `zamboni-quartz-19`. On main/staging (pre-fix): the model says it has no such passphrase in its context, meaning the message never reached it. See the screenshots above for a captured run of both against staging (before) and this PR's preview (after).
4. Quick local check: run `pnpm --filter @bike4mind/utils test -- utils.test.ts`. All tests pass, including the new suite `array-content fabMessages that carry no image block`. Reverting just `utils.ts` to `origin/main` and re-running fails that suite, including one test that throws a `TypeError` before the fix.

## Additional Information
N/A

## Developer Notes (Optional)
N/A

## Checklist

- [ ] I have added the new AdminSettings to Staging, prod and the hydration script (if applicable)
- [ ] I have made the UI/UX feel good (toasters, size, responsive, etc)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings
- [ ] If adding/modifying telemetry fields: updated [telemetry data classification](../docs-site/docs/security/telemetry-data-classification.md) doc

### CODE REVIEW

File: b4m-core/utils/src/llm/utils.ts [hunk 1]
```diff
@@ -244,6 +244,13 @@ const urlDerivedMessages = new WeakSet<IMessage>();
  */
 const IMAGE_TOKEN_ESTIMATE = 1600;
 
+/**
+ * Block types a caller-supplied user-role message can deliver on its own. An allowlist rather than
+ * a denylist, so a block type added to MessageContentTypes later is dropped loudly by the
+ * classifier below instead of forwarded to a provider that rejects it.
+ */
+const DELIVERABLE_USER_BLOCK_TYPES = new Set<string>(['text']);
```
Comment (onoya) [inline b4m-core/utils/src/llm/utils.ts:252]: Good call making this an allowlist rather than a denylist - if `MessageContentTypes` grows a new block type later, it fails loud (dropped with a warning) here instead of silently forwarding something a provider rejects. Nice given this file's history of silent-drop bugs.

File: b4m-core/utils/src/llm/utils.ts [hunk 2]
```diff
@@ -264,7 +271,11 @@ const estimateTokenLength = (text: string): number => {
   return Math.ceil(text.length / CHARS_PER_TOKEN);
 };
 
-const isImageBlock = (obj: { type?: string }): boolean => obj.type === 'image' || obj.type === 'image_url';
+// Array content is caller-controlled (extraContextMessages accepts z.array(z.any())), so a block can
+// be null/undefined/not-an-object at runtime despite the static type; every caller here iterates raw
+// message.content, so the null-safety has to live in this one shared helper rather than at each call site.
+const isImageBlock = (obj: { type?: string } | null | undefined): boolean =>
+  obj?.type === 'image' || obj?.type === 'image_url';
```
Comment (onoya) [inline b4m-core/utils/src/llm/utils.ts:278]: Good null-safety, and it's justified by the actual schema rather than defensive-for-its-own-sake - `extraContextMessages` accepts `z.array(z.any())`, so a block really can be `null`/untyped at runtime despite the static `{ type?: string }` type. Covered by the new "does not throw on a null content block" test.

File: b4m-core/utils/src/llm/utils.ts [hunk 3]
```diff
@@ -2126,9 +2137,76 @@ export async function buildAndSortMessages(
...
+    // as an image message rather than being caught by the undeliverable-block bucket below.
+    if (blocks.some(isImageBlock)) {
+      imageMessages.push(message);
+      return;
+    }
+    if (blocks.every(block => DELIVERABLE_USER_BLOCK_TYPES.has((block as { type?: string })?.type ?? ''))) {
+      nonImageMessages.push(message);
+      return;
+    }
+    undeliverableBlockMessages.push({ index, message });
```
Comment (onoya) [inline b4m-core/utils/src/llm/utils.ts:2176]: Non-blocking design note: a message mixing a deliverable `text` block with an undeliverable one (e.g. `text` + `tool_result`) gets dropped whole here rather than extracting the `text` block and dropping only the `tool_result`. That's intentional and tested ("drops a mixed text+tool_result message whole, not just the tool_result block"), and it's no worse than the old silent-drop - just flagging that legitimate text in a mixed array is still lost. Could be worth a follow-up if that combination turns out to matter in practice.

**Review by onoya — APPROVED** [review]
LGTM! 🚀

Real bug, cleanly fixed. The old code split fabMessages into two independent positive filters (image / non-image), which left a gap: a `role: 'user'` array-content message with no image block - text-only, or a `tool_use`/`tool_result`/`thinking` block - matched neither filter and vanished with no trace. Confirmed this is reachable today, not hypothetical: `extraContextMessages` in `b4m-core/common/src/llm.ts` validates array content as `z.array(z.any())`, so this shape is a normal accepted API input.

Traced the new classification pass through every downstream consumer of `nonImageMessages`/`imageMessages` (`isAttachment`, `attachmentLabel`, token counting, final assembly, `ensureToolPairingIntegrity`) and each one either already handled array content generically or is correctly guarded by this PR - e.g. `isAttachment` now requires string content, so a newly-admitted text-only array message can't be misreported as a dropped file attachment. 10 new tests cover the actual crash this fixes (untyped/null blocks) plus the loud-drop-with-warning path, and the warnings stay content-free.

A few things worth calling out inline below - none blocking.

**Review by ken-b4m — APPROVED** [review]
## PR Review: fix(llm-utils): stop silently dropping array-content user messages with no image block

**CI:** green. `awaiting testing` label. Skipped: `E2E Tests` (label-gated) plus build legs.
**Mergeable:** yes. Requirements source: #1113.
**Lenses:** none — 270 additions in two files, one classification pass. Inline.
**Executed:** yes — 174/174.

---

### Overall Assessment: **APPROVE** — replaces a gap between two filters with an exhaustive classification, and nothing can fall through it now.

The bug is the classic shape: two *positive* filters (`role === 'user' && !Array.isArray(content)` for one bucket, an image check for the other) with no assertion that they cover the space. A `role: 'user'` message whose content is an array carrying no image block — text-only, or a `tool_use`/`tool_result`/`thinking` block — matched neither and vanished with no trace. And it is reachable from an accepted API input, since `extraContextMessages` takes `z.array(z.any())`.

**The replacement really is exhaustive**, which is the thing worth checking about a "one pass, every message lands somewhere" claim. Walking it: `system` returns early (handled by `systemCandidates`), non-`user` goes to `unassignable` **with a warning**, non-array goes to `nonImageMessages`, an empty array goes to `unassignable` **with a warning**, any image block wins (checked first, so a block mixing an image with a `tool_result` still delivers as an image rather than being caught below), all-deliverable blocks go to `nonImageMessages`, and everything else goes to `undeliverable` **with a warning**. Every branch either assigns or reports. There is no path left that drops a message in silence, which was the actual defect — not that some messages are undeliverable, but that nobody could tell.

**The allowlist is the right polarity and the comment says why**: `DELIVERABLE_USER_BLOCK_TYPES` is a `Set(['text'])` rather than a denylist, so a block type added to `MessageContentTypes` later is dropped *loudly* by the classifier instead of forwarded to a provider that rejects it. That is the conservative direction for a boundary that accepts `z.any()`.

**The warnings explain themselves.** Rather than "dropped 2 messages", they say a `tool_result` supplied here has no `tool_use` to pair with — replayed history already pairs every `tool_use` it emits — so it would be stripped as an orphan after assembly, and a `thinking` block is assistant-only and the provider rejects it on a user turn. That is the difference between a log line someone silences and one that resolves the ticket.

**And the labels are content-free.** Block types, position and an estimated token count only — never payload text — explicitly matching the convention `attachmentLabel` already followed. On a path whose whole purpose is carrying caller-supplied context, that restraint matters: the natural implementation would have logged the dropped content, which is exactly the material you least want in a log sink.

The `attachmentLabel` fix is a small thing worth noting too — array content routed here has no filename header, so calling it "attachment N" would misreport a context block as a file; it now reports `context blocks [types]` instead.

---

### No findings

Nothing at P1, P2 or P3. The null-safety hardening on `isImageBlock` (`obj?.type`) is correctly placed in the one shared helper rather than at each call site, with the reason stated — content is caller-controlled, so a block can be null at runtime despite the static type.

---

### What's done well

The evidence in the PR body is the strongest form available for a silent-drop bug: the same request with the same passphrase against staging and against the preview, showing the model denying knowledge of it before and echoing it back after. A silent drop cannot be demonstrated by a test failing — the old code was "working" by its own lights — so proving delivery end-to-end against a live model is the right call.

Fixing the adjacent cases in the same pass (non-user/non-system messages, and the `.startsWith('image')` that could throw on a malformed block) is also right: they are the same class of defect in the same function, and leaving them would mean the next person rediscovers the gap from a different direction.

**Comment by gsreyson** [comment]
# B4M-1635 Test Results

- **PR:** [Bike4Mind/bike4mind #1635](https://github.com/Bike4Mind/bike4mind/pull/1635) — fix(llm-utils): stop silently dropping array-content user messages with no image block
- **Test site:** https://app.pr1635.preview.bike4mind.com
- **Tested by:** Claude Code (test-bike4mind-pr skill)
- **Date:** 2026-08-12
- **Users:** admin (qa-admin-e2e@test.com). The normal user (qa-user-e2e@test.com) was logged in successfully but is stuck behind the first-login policy-acceptance gate on this preview, which 403s every authenticated endpoint. Accepting policies is a consent action, so it was not clicked; the admin account had already accepted, and the guide does not require a specific role.

## Verdict

**✅ PASS** — 4/4 guide steps PASS, plus 2 extra checks I added (negative control and the documented drop-path) also PASS.

The fix is confirmed live with a real before/after against two environments: the identical request returns the passphrase on the PR preview and fails to on staging (pre-fix).

**No failures, and no new bugs introduced by this PR.** The items in "Bugs / observations" are split so nothing there reads as fallout from this change: section A is the bug this PR fixes (reproduced live on staging), section B is new bugs from this PR (none), section C is pre-existing issues unrelated to it, and section D is what I could not verify.

One process note: this run has **no screenshots** — see Environment notes.

## Steps

Status emoji: ✅ PASS · ❌ FAIL · ⚠️ BLOCKED · ➖ NOT TESTED.

| # | Step | Expected result | Actual result | Status | Evidence |
|---|------|-----------------|---------------|--------|----------|
| 1 | Authenticate against the preview and `POST /api/ai/llm` with the guide's exact body: array-content `extraContextMessages` carrying `zamboni-quartz-19`, model `us.anthropic.claude-haiku-4-5-20251001-v1:0` | HTTP 200 with a `quest.id` | HTTP 200, `quest.id = 6a7be3604d6f1bc2d8ba7202` | ✅ PASS | `req-fix.json`, `resp-fix.json` |
| 2 | Poll `GET /api/quests/{id}` until `replies` is non-empty | Reply eventually arrives | Reply arrived within the poll window | ✅ PASS | `quest-fix.json` |
| 3 | Check the reply content | Reply contains `zamboni-quartz-19` | Reply was exactly `zamboni-quartz-19` — nothing else. The context message reached the model. | ✅ PASS | `quest-fix.json` |
| 4 | Confirm the pre-fix behavior on main/staging | Model says it has no such passphrase | Ran the **identical** request body against `app.staging.bike4mind.com` (pre-fix): *"I don't see an archive passphrase in my context..."* — the message never reached the model, reproducing the bug | ✅ PASS | `quest-staging.json` |
| 5 | *(added)* Negative control on the preview: same prompt, **no** `extraContextMessages` | Model should have no passphrase — proves step 3 is attributable to the fixed field, not to leaked history or a lucky guess | *"I don't see an archive passphrase mentioned in my context..."* | ✅ PASS | `req-control.json`, `quest-control.json` |
| 6 | *(added)* Drop-path: array content mixing a `text` block **and** a `tool_result` block | Per the PR, the message is dropped **whole** (not partially delivered) and the request still succeeds rather than erroring | HTTP 200, normal reply, and the model did **not** have the passphrase — the whole message was dropped as designed, no 500, no partial text delivery | ✅ PASS | `req-toolresult.json`, `quest-toolresult.json` |
| 7 | Local: `pnpm --filter @bike4mind/utils test -- utils.test.ts` | All pass, including the new array-content suite | 41 files / 813 tests passed. The new suite `array-content fabMessages that carry no image block` is present at `b4m-core/utils/src/llm/utils.test.ts:2519` and all 10 of its tests pass (verified individually with `--reporter=verbose`) | ✅ PASS | see Test output below |

### Local test output (step 7)

```
Test Files  41 passed (41)
     Tests  813 passed (813)
```

All 10 tests in the new suite, run verbosely:

```
✓ delivers a text-only array-content message to the model (fails against the old two-filter split)
✓ drops a tool_result-bearing array-content message with a warning naming what and why
✓ drops a mixed text+tool_result message whole, not just the tool_result block
✓ keeps image priority when an image block shares an array with a tool_result block
✓ drops empty array content with a warning, never as a blank message
✓ drops a thinking-only array-content message with a named warning
✓ delivers two qualifying text-only array-content messages in order
✓ ignores a non-user-role fabMessage with a named warning (the split-filter hole beyond user array content)
✓ does not throw on an untyped content block
✓ does not throw on a null content block
```

Note: the guide's `-- utils.test.ts` argument does not actually filter vitest to that one file (it runs the whole package suite). Harmless — everything passes either way — but if the intent was to run just that file, `npx vitest run src/llm/utils.test.ts` is the command that does it.

## Screenshots

**None.** This run captured no PNGs. Two independent reasons:

1. The Playwright MCP browser profile (`ms-playwright-mcp/mcp-chrome-79a444b`) was locked by a Chrome process left running from a previous session (PID 70663, launched Aug 11 20:57, parked at `about:blank`). Every Playwright call returned `Browser is already in use ... use --isolated`. The skill's rules forbid closing a browser this run did not open, so it was left alone.
2. The feature under test has **no UI affordance** — the guide itself says only a programmatic caller would send this shape. There is nothing to photograph; the evidence is the request/response pair.

Saved instead, as machine-checkable evidence, are the raw API transcripts:

| File | What it shows |
|------|---------------|
| `req-fix.json` | The exact request body from the guide (array-content `extraContextMessages` with the passphrase) [req-fix.json](https://github.com/user-attachments/files/30965957/req-fix.json) |
| `resp-fix.json` | The `POST /api/ai/llm` 200 response, including the `quest.id` and the resolved model metadata [resp-fix.json](https://github.com/user-attachments/files/30965956/resp-fix.json) |
| `quest-fix.json` | **The key artifact** — the completed quest on the PR preview, `replies: ["zamboni-quartz-19"]` [quest-fix.json](https://github.com/user-attachments/files/30965955/quest-fix.json) |
| `quest-staging.json` | The identical request on staging (pre-fix): model reports no passphrase — the bug, reproduced [quest-staging.json](https://github.com/user-attachments/files/30965950/quest-staging.json) |
| `req-control.json` / `quest-control.json` | Negative control on the preview with the field omitted — model reports no passphrase [req-control.json](https://github.com/user-attachments/files/30965954/req-control.json) |
| `req-toolresult.json` / `quest-toolresult.json` | The `tool_result`-bearing message: dropped whole, request still 200 [req-toolresult.json](https://github.com/user-attachments/files/30965952/req-toolresult.json) |
| `models.json` | `GET /api/models` output, used to pick an available text model [models.json](https://github.com/user-attachments/files/30965949/models.json) |

To re-capture this with UI screenshots later, close the stale Chrome (or run Playwright MCP with `--isolated`) — but note it would only produce a login screenshot, since the field has no UI.

## Bugs / observations

### A. The bug this PR fixes — confirmed real, confirmed fixed

- **Reproduced on staging (pre-fix):** an array-content `extraContextMessages` entry with no image block was silently discarded. Identical request body, identical model, only the environment differs: staging says it has no passphrase, the PR preview echoes `zamboni-quartz-19`. This is the defect the PR describes, observed live rather than inferred from the diff.
- **Fixed on the preview:** the message now reaches the model. Verified with a negative control (field omitted -> no passphrase), so the delivery is attributable to this change and not to conversation history or a lucky guess.

### B. New bugs introduced by this PR

- **None found.** Nothing in the change under test misbehaved across the 6 live checks.

### C. Pre-existing issues and observations — NOT caused by this PR

None of these are regressions from #1635; they were present before it and are unaffected by it. Listed so they are not mistaken for fallout from this change.

- **Policy-acceptance gate blocks the normal QA user on a fresh preview.** `qa-user-e2e@test.com` has not accepted policies here, so every authenticated API call for that account returns `403 {"policyAcceptanceRequired": true}` — including `GET /api/models`. This is a *preview-env data state*, not a code defect, and it is unrelated to LLM message handling. Any future guide step needing the normal user on a fresh preview will be blocked until someone accepts. Flagged rather than acted on, since accepting terms on the user's behalf is a consent decision. The `403` body itself is well-formed and its `error_description` explains the remedy, so the gate is behaving correctly.
- **The guide's local-test command does not filter the way it reads.** `pnpm --filter @bike4mind/utils test -- utils.test.ts` runs the whole package suite rather than that one file. Harmless — everything passes either way — but `npx vitest run src/llm/utils.test.ts` is the command that does what the guide appears to intend. This is a documentation nit in the PR body, not a product bug.
- **Stale Playwright Chrome held the browser profile lock** (PID 70663, launched Aug 11, parked at `about:blank`), which is why this run has no screenshots. A local tooling condition, entirely outside the product.

### D. Verification limits (things I could not check, rather than things that are wrong)

- The PR states a warning is logged naming what was dropped and why. Server-side logs are not readable from the client, so that claim is verified by unit test (step 7) rather than observed in the live environment. What I *could* confirm live is the user-visible half: the drop is graceful — a `tool_result`-bearing message returns a normal 200 with a normal answer instead of erroring or partially delivering.
